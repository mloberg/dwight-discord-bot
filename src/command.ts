import { SlashCommandBuilder, SlashCommandIntegerOption } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';

type Handler = (command: CommandInteraction) => Promise<void>;

export class CommandBuilder extends SlashCommandBuilder {
    constructor(readonly handler: Handler) {
        super();
    }
}

export const rollOption = (dice: string): SlashCommandIntegerOption =>
    new SlashCommandIntegerOption().setName('roll').setDescription(`Value of a ${dice} roll.`);
