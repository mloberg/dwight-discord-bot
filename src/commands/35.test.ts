import { Client, Guild, Message, TextChannel } from 'discord.js';

import { FriendlyError } from '../error';
import command, { conversion } from './35';

const mocks = {
    reply: jest.fn(),
};

jest.mock('discord.js', () => ({
    Client: jest.fn(),
    Guild: jest.fn(),
    TextChannel: jest.fn(),
    Collection: jest.fn(),
    Message: jest.fn().mockImplementation(() => {
        return {
            reply: mocks.reply,
        };
    }),
}));

describe('_3.5 configuration', () => {
    it('should have basic command infomation', () => {
        expect(command.name).toEqual('3.5');
        expect(command.description).toEqual('3.5 to 5th edition skills conversion');
        expect(command.usage).toEqual('SKILL');
    });

    it('should have aliases', () => {
        expect(command.alias).toContain('35');
    });
});

describe('_3.5', () => {
    let message: Message;

    beforeEach(() => {
        mocks.reply.mockClear();
        mocks.reply.mockReturnThis();

        const client = new Client();
        const guild = new Guild(client, {});
        const channel = new TextChannel(guild, {});
        message = new Message(client, {}, channel);
    });

    it.each([
        [['use', 'rope'], 'DEX (Acrobatics)'],
        [['Use', 'Magic', 'Device'], 'INT (Arcana)'],
        [['Ride'], 'WIS (Animal Handling) or DEX (Acrobatics)'],
    ])('returns a comprable 5e skill for %s', async (search, result) => {
        await command.run(message, { $0: '3.5', _: search });

        expect(mocks.reply).toHaveBeenCalledTimes(1);
        expect(mocks.reply).toHaveBeenCalledWith(result);
    });

    it('has the correct conversions', () => {
        expect(conversion).toMatchSnapshot();
    });

    it('throws an error on invalid skill', async () => {
        try {
            await command.run(message, { $0: '3.5', _: ['foo'] });

            fail('expected error to be thrown');
        } catch (err) {
            expect(err instanceof FriendlyError).toBe(true);
            expect(err.message).toEqual('I couldn\'t find skill "foo".');
        }
    });
});
