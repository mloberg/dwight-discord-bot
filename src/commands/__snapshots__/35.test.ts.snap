// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/35 is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Convert 3.5 skill to 5e",
    "name": "35",
    "options": Array [
      Object {
        "autocomplete": undefined,
        "choices": undefined,
        "description": "3.5 skill",
        "name": "skill",
        "required": true,
        "type": 3,
      },
    ],
  },
  "handle": [Function],
}
`;
